src/plugins/intel_gpu/src/graph/impls/ocl_v2/moe_expert_opt.cpp

```diff
@@ -0,0 +1,1215 @@
+    }
+
+    [[nodiscard]] std::unique_ptr<primitive_impl> clone() const override {
+        auto moe = make_deep_copy<MoeExpertOptImpl>(this);
+        moe->_dnnl_weights = _dnnl_weights;
+        moe->_hidden_size = _hidden_size;
+        moe->_intermediate_size = _intermediate_size;
+        moe->_group_size = _group_size;
+        moe->_up_use_cl = _up_use_cl;
+        moe->_down_use_cl = _down_use_cl;
+        return moe;
+    }
+
+    std::vector<BufferDescriptor> get_internal_buffer_descs(const kernel_impl_params& params) const override {
+        auto moe = params.typed_desc<moe_expert>();
+        const auto& config = moe->_config;
+        int max_topk = static_cast<int>(config.topk);
+        int expert_num = static_cast<int>(config.expert_num);
+
+        auto hidden_states_layout = params.input_layouts[0];
+        auto batch = static_cast<int>(hidden_states_layout.get_shape()[0]);
+        auto data_type = hidden_states_layout.data_type;
+
+        std::vector<BufferDescriptor> internal_buffers;
+        // softmax+topk
+        layout layout_topk_id(ov::PartialShape{batch, max_topk}, data_types::u32, cldnn::format::bfyx);
+        layout layout_topk_weights(ov::PartialShape{batch, max_topk}, data_type, cldnn::format::bfyx);
+        internal_buffers.emplace_back(layout_topk_id, false);       // topk_id
+        internal_buffers.emplace_back(layout_topk_weights, false);  // topk_weights
+        // fast single batch: scratch.up = up(x) * silu(gate(x)); scratch.y = down(scratch.up) * weight[expert_no]
+        layout layout_gateup_out(ov::PartialShape{batch, static_cast<int>(config.intermediate_size)}, data_type, cldnn::format::bfyx);
+        layout layout_down_out(ov::PartialShape{batch, static_cast<int>(config.hidden_size)}, data_type, cldnn::format::bfyx);
+        internal_buffers.emplace_back(layout_gateup_out, false);  // up
+        internal_buffers.emplace_back(layout_down_out, false);    // y
+        // onednn: scratch.x, scratch.routing_weights = gather(x, ...)
+        //         scratch.up = up(scratch.x)
+        //         scratch.gate = gate(scratch.x) * scratch.up
+        //         scratch.y = down(scratch.gate) * routing_weights
+        internal_buffers.emplace_back(layout_down_out, false);  // x, scratch.x has same layout with down output
+        layout routing_layout(ov::PartialShape{batch * max_topk}, data_type, cldnn::format::bfyx);
+        internal_buffers.emplace_back(layout_down_out, false);    // routing_weights
+        internal_buffers.emplace_back(layout_gateup_out, false);  // gate, scratch.gate has same layout with up
+        // expert masks for gpu
+        layout index_layout(ov::PartialShape{batch}, ov::element::i32, cldnn::format::bfyx);
+        for (int i = 0; i < expert_num; i++) {
+            internal_buffers.emplace_back(index_layout, true);  // batch
+            internal_buffers.emplace_back(index_layout, true);  // topk
+        }
```

Comment:
I think it would be better to use kernel's scalar arguments for this data instead of internal_buffers, as scalars should have better performance than usm_host allocations

Reply:
`internal_buffers` is a vector of `cldnn::BufferDescriptor`, it will not be sent to kernel.

Reply:
For sure, my point is that it allocates single-element buffers for the batch and top_k values. I think we shouldn't use buffers for passing these values at all - we can just pass scalar values to the kernel instead